cpp/docs/DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,977 @@
+                           not std::is_floating_point<ColumnType>::value>* = nullptr> 
+void operator()() { std::cout << "neither integral nor floating point\n"; }
+
+template <typename ColumnType,
+          std::enable_if_t<std::is_integral<ColumnType>::value>* = nullptr>
+void operator()() { std::cout << "integral\n"; }
+
+template < typename ColumnType,
+           std::enable_if_t<std::is_floating_point<ColumnType>::value>* = nullptr> 
+void operator()() { std::cout << "floating point\n"; }
+};
+```
+
+For more info on SFINAE with `std::enable_if`, [see this post](https://eli.thegreenplace.net/2014/sfinae-and-enable_if).
+
+There are a number of traits defined in `include/cudf/utilities/traits.hpp` that are useful for 
+partial specialization of dispatched function objects. For example `is_numeric<T>()` can be used to 
+specialize for any numeric type.
+
+# Nested Data Types
+
+libcudf supports a number of nested data types, including strings, lists, and structs. A string is 
+simply a character string, but a column of strings may have a different-length string in each row.
+A list is a list of elements of any type, so a column of lists of integers has rows with a list of 
+integers, possibly of a different length, in each row. In a colum of structs, each row is a 
+structure comprising one or more fields. These fields are stored in structure-of-arrays format, so 
+that the column of structs has a nested column for each field of the structure. 
+
+As the heading implies, these column types may be nested arbitrarily. One may create a column of 
+lists of structs, where the fields of the struct may be of any type, including strings, lists and 
+structs. Thinking about deeply nested data types can be confusing for column-based data, even with 
+experience. Therefore it is important to carefully write algorithms, and to test and document them
+well.
+
+## List columns
+
+In order to represent variable-width elements, libcudf columns contain a vector of child columns.
+For list columns, the parent column's type is `LIST` and contains no data, but its size represents
+the number of lists in the column, and its null mask represents the validity of each list element.
+The parent has two children. 
+
+1. A non-nullable column of `INT32` elements that indicates the offset to the beginning of each list
+   in a dense column of elements.
+2. A column containing the actual data and optional null mask for all elements of all the lists 
+   packed together.
+   
+With this representation, `data[offsets[i]]` is the first element of list `i`, and the size of list
+`i` is given by `offsets[i+1] - offsets[i]`.
+
+Note that the data may be of any type, and therefore the data column may itself be a nested column
+of any type. Note also that not only is each list nullable (using the null mask of the parent), but
+each list element may be nullable. So you may have a lists column with null row 3, and also null
+element 2 of row 4.
+
```

Comment:
A key point with list columns is:
- The actual underlying data for a lists column is always bundled together into a single leaf column at the very bottom of the hierarchy (ignoring structs, which conceptually "reset" the root of the hierarchy), regardless of the level of nesting.  So a `List<List<List<List<int>>>>>` column still just has a single int column at the very bottom.  

A semi-visual representation might be useful here:

```
   lists_column = { {{{1, 2}, {3, 4}}, NULL}, {{{10, 20}, {30, 40}}, {{50, 60, 70}, {0}}} }

   List<List<List<int>>>  (2 rows):
   Length : 2
   Offsets : 0, 2, 4
   Children :
      List<List<int>>:
      Length : 4
      Offsets : 0, 2, 2, 4, 6
      Null count: 1
        1101
      Children :
        List<int>:
        Length : 6
        Offsets : 0, 2, 4, 6, 8, 11, 12
        Children :
          Column of ints
          1, 2, 3, 4, 10, 20, 30, 40, 50, 60, 70, 0
```

Reply:
A link to the arrow docs would be good too:

https://arrow.apache.org/docs/format/Columnar.html?highlight=nested%20types#physical-memory-layout

Our list implementation is specifically the "Variable-size List" physical layout.

Reply:
Added all of this. Thanks.